notebooks/ch16/classification_sgd.md

```diff
@@ -0,0 +1,187 @@
+
+
+```python
+# HIDDEN
+import numpy as np
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
+%matplotlib inline
+import ipywidgets as widgets
+from ipywidgets import interact, interactive, fixed, interact_manual
+import nbinteract as nbi
+
+from sklearn.linear_model import SGDClassifier
+from sklearn.datasets import load_iris
+from sklearn.feature_extraction import DictVectorizer
+from sklearn.model_selection import train_test_split
+
+sns.set()
+sns.set_context('talk')
+np.set_printoptions(threshold=20, precision=2, suppress=True)
+pd.options.display.max_rows = 7
+pd.options.display.max_columns = 8
+pd.set_option('precision', 2)
+np.random.seed(42)
+```
+
+## Gradient Descent in Logistic Regression
+
+Previously, we covered batch gradient descent, an algorithm that iteratively updates $\hat\theta$ to find the cost-minimizing parameters. We also discussed stochastic gradient descent and mini-batch gradient descent, methods that take advantage of statistical theory and parallelized hardware to decrease the time spent training the gradient descent algorithm. In this section, we will apply these concepts specifically to logistic regression and walk through examples using scikit-learn functions.
+
+### Batch Gradient Descent
+
+The general update formula for batch gradient descent is given by:
+
+$$
+\hat\theta_{t+1} = \hat\theta_t - \alpha \cdot \nabla_\hat\theta L(\hat\theta, X, y)
+$$
+
+In logistic regression, we use the cross entropy cost as our cost function. $\nabla_\hat\theta L(\hat\theta, X, y)$ is then the gradient of the cross entropy cost, $-\frac{1}{n}\sum_{i=1}^n(y_i - \sigma_i)X_i$; plugging this in allows us to find the gradient descent algorithm specific to logistic regression. Letting $ \sigma_i = f_\hat\theta(X_i) = \sigma(X_i \cdot \hat \theta) $, this becomes:
+
+$$
+\begin{align}
+\hat\theta_{t+1} &= \hat\theta_t - \alpha \cdot \left(- \frac{1}{n} \sum_{i=1}^{n} \left(y_i - \sigma_i\right) X_i \right) \\
+&= \hat\theta_t + \alpha \cdot \left(\frac{1}{n} \sum_{i=1}^{n} \left(y_i - \sigma_i\right) X_i \right)
+\end{align}
+$$
+
+### Stochastic Gradient Descent
+
+The general update formula is below, where $l(\hat\theta, X, y)$ is the loss function for a single data point:
```

Comment:
I may be interpreting this wrong, but shouldn't it be X_i and y_i for a single data point instead of X and y?

Reply:
You're right, that's a typo!